src/Data/Nat/Primality/Factorisation.agda

```diff
@@ -0,0 +1,192 @@
+-- Finding a factorisation
+
+-- this builds up three important things:
+-- * a proof that every number we've gotten to so far has increasingly higher
+--   possible least prime factor, so we don't have to repeat smaller factores
+--   over and over (this is the "m" and "rough" parameters)
+-- * a witness that this limit is getting closer to the number of interest, in a
+--   way that helps the termination checker (the "k" and "eq" parameters)
+-- * a proof that we can factorise any smaller number, which is useful when we
+--   encounter a factor, as we can then divide by that factor and continue from
+--   there without termination issues
+
+private
+  pattern 2+ n = suc (suc n)
+  pattern 2≤2+n = s≤s (s≤s z≤n)
+  pattern 1<2+n = 2≤2+n
+
+factorise : ∀ n → .{{NonZero n}} → PrimeFactorisation n
+factorise 1 = record
+  { factors = []
+  ; isFactorisation = refl
+  ; factorsPrime = []
+  }
+factorise n₀@(2+ _) = build [ <-recBuilder ⊗ <-recBuilder ] P facRec (n₀ , suc n₀ ∸ 4) 2-rough refl
+  where
+  P : ℕ × ℕ → Set
+  P (n , k) = ∀ {m} → .{{NonTrivial n}} → .{{NonTrivial m}} → m Rough n → suc n ∸ m * m ≡ k → PrimeFactorisation n
+
+  facRec : ∀ n×k → (<-Rec ⊗ <-Rec) P n×k → P n×k
+  -- Case 1: m * m > n, ∴ Prime n
+  facRec (n , zero) _ rough eq = record
+    { factors = n ∷ []
+    ; isFactorisation = *-identityʳ n
+    ; factorsPrime = rough∧>square⇒prime rough (m∸n≡0⇒m≤n eq) ∷ []
+    }
+  facRec (n@(2+ _) , suc k) (recFactor , recQuotient) {m@(2+ _)} rough eq with m ∣? n
+  -- Case 2: m ∤ n, try larger m, reducing k accordingly
+  ... | no m∤n = recFactor (≤-<-trans (m∸n≤m k (2 * m)) (n<1+n k)) {suc m} (∤⇒rough-suc m∤n rough) $ begin
+    suc n ∸ (suc m + m * suc m)   ≡⟨ cong (λ # → suc n ∸ (suc m + #)) (*-suc m m) ⟩
+    suc n ∸ (suc m + (m + m * m)) ≡˘⟨ cong (suc n ∸_) (+-assoc (suc m) m (m * m)) ⟩
+    suc n ∸ (suc (m + m) + m * m) ≡⟨ cong (suc n ∸_) (+-comm (suc (m + m)) (m * m)) ⟩
+    suc n ∸ (m * m + suc (m + m)) ≡˘⟨ cong (λ # → suc n ∸ (m * m + suc (m + #))) (+-identityʳ m) ⟩
+    suc n ∸ (m * m + suc (2 * m)) ≡˘⟨ ∸-+-assoc (suc n) (m * m) (suc (2 * m)) ⟩
+    (suc n ∸ m * m) ∸ suc (2 * m) ≡⟨ cong (_∸ suc (2 * m)) eq ⟩
+    suc k ∸ suc (2 * m)           ∎
+    where open ≡-Reasoning
+  -- Case 3: m ∣ n, record m and recurse on the quotient
+  ... | yes m∣n = record
+    { factors = m ∷ ps
+    ; isFactorisation = m*Πps≡n
+    ; factorsPrime = rough∧∣⇒prime rough m∣n ∷ primes
+    }
+    where
+      m<n : m < n
+      m<n = begin-strict
+        m            <⟨ s≤s (≤-trans (m≤n+m m _) (+-monoʳ-≤ _ (m≤m+n m _))) ⟩
+        pred (m * m) <⟨ s<s⁻¹ (m∸n≢0⇒n<m λ eq′ → 0≢1+n (trans (sym eq′) eq)) ⟩
+        n            ∎
+        where open ≤-Reasoning
+      q = quotient m∣n
+      instance _  = n>1⇒nonTrivial (quotient>1 m∣n m<n)
+      factorisation[q] : PrimeFactorisation q
+      factorisation[q] = recQuotient (quotient-< m∣n) (suc q ∸ m * m) (rough∧∣⇒rough rough (quotient-∣ m∣n)) refl
```

Comment:
I'll also look at the call structure overall, in case this can be simplified too (eg wrt the `suc n ∸ m * m ≡ k` variant relation...)